### PR TITLE
Fixed device.disable_and_clear_program_cache() not working

### PR DESCRIPTION
**Pull request by dmakoviichuk-tt**
### Ticket
Not ticket.

### Problem description
In the device_operator.cpp we have check if cache is enabled, and if not we call `create_or_get_program_from_cache`. But in hpp file `create_or_get_program_from_cache` we always adding to cache then reading from it even if it is disabled. And inside of the program cache we call `try_emplace` as result we are reusing old cached value.

This is attempt to fix, but it might not work. Previously we put copy of the cached program. No I am trying to avoid it.

### What's changed
Applied small update to the Program Cache API.
Fixed `create_or_get_program_from_cache`.
Another one possible solution is to try call `operator[]` instead of `try_emplace`.
But as for me adding some value to the map and reading it in the next line immediately doesn't look right.

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt_metal/impl/device/program_cache.hpp [hunk 1]
```diff
@@ -18,7 +18,7 @@ namespace detail {
 // Generic Program Cache: This data structure is tied to a device handle and can store generic program types from
 // TT-Metal and TT-Eager using tt::stl::concepts::unique_any.
 struct ProgramCache {
-    inline bool contains(uint64_t program_hash) { return this->cache_.count(program_hash) > 0; }
+    inline bool contains(uint64_t program_hash) { return is_enabled() ? this->cache_.count(program_hash) > 0 : false; }
 
     template <typename T>
     inline T& get(uint64_t program_hash) {
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/device/program_cache.hpp:24]: Did you consider disabling this too?

**Review by ayerofieiev-tt — APPROVED** [review]

File: tt_metal/impl/device/program_cache.hpp (hunk 1 above)
Comment (dmakoviichuk-tt) [reply tt_metal/impl/device/program_cache.hpp:24]: Have no idea what to return? TT_FATAL then?

**Review by eyonland — APPROVED** [review]

**Comment by aliuTT** [comment]
Might not be the best platform to ask this question, but what is the alignment `static constexpr auto ALIGNMENT = 32;` requirement? If it end's up being the same as the hardware or PCI alignment constraints, maybe it's better to pull that from `tt_metal/hostdevcommon/common_runtime_address_map.h`

**Comment by dmakoviichuk-tt** [comment]
@aliuTT I think its half a random number to make sure all programs have at least this or lower alignment.
@arakhmati might know more.

**Comment by dmakoviichuk-tt** [comment]
Closing this ticket.
After additional review decided to go another way.
We must use shared_pointer instead of reference.

**Pull request closed without merging**